arangod/Aql/Condition.cpp

```diff
@@ -1914,29 +1974,60 @@ AstNode* Condition::transformNodePostorder(
       //  a   c      b   c
       //
 
-      auto newOperator = _ast->createNode(NODE_TYPE_OPERATOR_NARY_OR);
-
-      std::vector<::PermutationState> clauses;
+      ::arangodb::containers::SmallVector<::PermutationState, 4> clauses;
       clauses.reserve(n);
 
+      size_t orMembers = 1;
       for (size_t i = 0; i < n; ++i) {
         auto sub = node->getMemberUnchecked(i);
+        size_t const subMembers =
+            (sub->type == NODE_TYPE_OPERATOR_NARY_OR ? sub->numMembers() : 1);
 
-        if (sub->type == NODE_TYPE_OPERATOR_NARY_OR) {
-          clauses.emplace_back(sub, sub->numMembers());
-        } else {
-          clauses.emplace_back(sub, 1);
-        }
+        clauses.emplace_back(sub, subMembers);
+        orMembers *= subMembers;
       }
+      TRI_ASSERT(clauses.size() == n);
+
+      // fetch maximum number of condition members from query options
+      size_t maxNumberOfConditionMembers =
+          _ast->query().queryOptions().maxDNFConditionMembers;
+
+      // check if there would be too many members in the DNF version of
+      // the condition. we may suffer from combinatorial explosion here.
+      if (orMembers >= maxNumberOfConditionMembers) {
+        // throw a special error about a too complex query condition.
+        // this will be handled by the callers, so that they will continue
+        // with a simplified version of the condition (which may not be
+        // usable for index lookups).
+        THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_FAILED,
+                                       "too complex query condition");
```

Comment:
If I read that correctly, you want to catch only this specific error in `transformCondition`. However, it seems to be used for other things as well - wouldn't it make sense to introduce a specific error code for this, or did I misunderstand the intention?

Reply:
Yes, the intention was to catch only this specific error. I reused TRI_ERROR_FAILED because it is not used often in other places (at least in AQL). I can introduce a specific error for this case to make it more future-proof.

Reply:
Addressed.